tests/subsys/net/lib/nrf_cloud/nrf_cloud_test/src/nrf_cloud_test.c

```diff
@@ -0,0 +1,654 @@
+	RESET_FAKE(nct_connect);
+	RESET_FAKE(nct_socket_get);
+	RESET_FAKE(nct_keepalive_time_left);
+	RESET_FAKE(nct_process);
+}
+
+/* This function runs after each completed test */
+static void run_after(void *fixture)
+{
+#if defined(CONFIG_NRF_CLOUD_FOTA)
+	nrf_cloud_fota_uninit_fake.custom_fake = fake_nrf_cloud_fota_uninit__succeeds;
+#endif
+	nct_disconnect_fake.custom_fake = fake_nct_disconnect__succeeds;
+
+	/* Uninit to reset the internal state to idle */
+	nrf_cloud_uninit();
+}
+
+ZTEST_SUITE(nrf_cloud_test, NULL, NULL, run_before, run_after, NULL);
+
+/* TEST NRF_CLOUD_INIT */
+
+/* The library can only be initialized when the internal state is
+ * idle, so the test verifies that nrf_cloud_init function will
+ * return with permission denied error (-EACCES) when not idle
+ */
+ZTEST(nrf_cloud_test, test_init_not_idle)
+{
+	struct nrf_cloud_init_param params = {
+		.event_handler = event_handler,
+		.client_id = NULL,
+	};
+
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	/* Call nrf_cloud_init twice with the first one
+	 * being successful and check if the second one
+	 * will fail since the state will not be idle
+	 */
+	nct_init_fake.custom_fake = fake_nct_init__succeeds;
+	nfsm_init_fake.custom_fake = fake_nfsm_init__succeeds;
+	nrf_cloud_codec_init_fake.custom_fake = fake_nrf_cloud_codec_init__succeeds;
+	nrf_cloud_init(&params);
+
+	int ret = nrf_cloud_init(&params);
+
+	zassert_not_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should not be in the idle state after initialization");
+
+	zassert_equal(-EACCES, ret, "return should be -EACCES when not idle");
+}
+
+/* Verify that nrf_cloud_init function will throw an error
+ * of invalid argument (-EINVAL) when given null params
+ */
+ZTEST(nrf_cloud_test, test_init_null_params)
+{
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	int ret = nrf_cloud_init(NULL);
+
+	zassert_equal(-EINVAL, ret, "return should be -EINVAL with NULL params");
+}
+
+/* Verify that nrf_cloud_init function will throw an
+ * error of invalid argument (-EINVAL) when a null
+ * event handler is passed in as part of its parameters
+ */
+ZTEST(nrf_cloud_test, test_init_null_handler)
+{
+	struct nrf_cloud_init_param params = {
+		.event_handler = NULL,
+		.client_id = NULL,
+	};
+
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	int ret = nrf_cloud_init(&params);
+
+	zassert_equal(-EINVAL, ret, "return should be -EINVAL with NULL event handler");
+}
+
+/* Verify that nrf_cloud_init function succeeds given valid internal
+ * state, event handler with full modem FOTA update is DISABLED.
+ * Other tests will not need FMFU since it is already tested here.
+ */
+ZTEST(nrf_cloud_test, test_init_success_no_fota)
```

Comment:
Also bear in mind that all of your other tests are running both with FMFU and without FMFU. Is that intended?

Reply:
> You'll want `Z_TEST_SKIP_IFDEF` I think

Yes, thanks for noting it out.

> I wonder if these two tests could be combined by just always setting the fmfu fake, regardless of whether FMFU is enabled? If not, I agree that this is the best approach. Though I would consider separating the FMFU-only tests to their own file. I think they can still be part of the `nrf_cloud_test` suite, it's just an organizational thing.

I'll think about this.

> Also bear in mind that all of your other tests are running both with FMFU and without FMFU. Is that intended?

Yes, it's intended since they do not use fmfu.

Reply:
They don't use fmfu, but if it is enabled, they still invoke it when they call init, right? Doesn't the `nrf_cloud_fota_fmfu_dev_set_fake.custom_fake` fake need to be set, or can it handle not being set at all?

Reply:
if there is no `param->fmfu_dev_inf` then it will log a warning. I guess that fake does not need to be set in that case.

Reply:
I combined the two functions into one